viewing_party/party.py

```diff
@@ -1,23 +1,276 @@
+            "genre": genre,
+            "rating": rating}
+        return new_movie
+    else:
+        new_movie = None
+        return new_movie
+
+# am i supposed to make this or is the test making it?
+# user_data = {
+#         "watched": [
+#             {
+#                 "title": "Title A",
+#                 "genre": "Horror",
+#                 "rating": 3.5
+
+#             }
+            
+#         ]
+#     }
+
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if title == movie["title"]:
+            user_data["watchlist"].remove(movie) 
+            user_data["watched"].append(movie)
+            return user_data
+
+    return user_data
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    total_ratings = 0
+    num_of_movies = len(user_data["watched"])
+    
+    if len(user_data["watched"]) < 1:
+        watched_avg_rating = 0.0 
+        return watched_avg_rating
+
+    for movie in user_data["watched"]:
+        total_ratings += movie["rating"]
+
+    watched_avg_rating = total_ratings / num_of_movies
+
+    return watched_avg_rating
+
+def get_most_watched_genre(user_data):
+    fantasy_count = 0
+    action_count = 0
+    intrigue_count = 0
+    
+    if len(user_data["watched"]) < 1:
+        most_watched_genre = None 
+        return most_watched_genre
+
+    for movie in user_data["watched"]:
+        if movie["genre"] == "Fantasy":
+            fantasy_count += 1
+        if movie["genre"] == "Action":
+            action_count += 1
+        if movie["genre"] == "Intrigue":
+            intrigue_count += 1
+
+    if fantasy_count > action_count and fantasy_count > intrigue_count:
+        most_watched_genre = "Fantasy"
+    if action_count > fantasy_count and action_count > intrigue_count:
+        most_watched_genre = "Action"
+    if intrigue_count > action_count and intrigue_count > fantasy_count:
+        most_watched_genre = "Intrigue"
+    return most_watched_genre
+
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    user_watched_list = []
+    friend_watched_list = []
+
+    for movie in user_data["watched"]:
+        user_watched_list.append(movie["title"])
+
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_watched_list.append(movie["title"])
+    
+    user_watched_set = set(user_watched_list)
+    friend_watched_set = set(friend_watched_list)
+    user_unique_watched_set = user_watched_set.difference(friend_watched_set)

```

Comment:
Great use of set operations. Since sets only allow immutable data, we had to use a unique(ish) immutable part of a move to stand in for the whole record.